plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -115,22 +121,28 @@ makePrisms ''ChainIndexTxOut
 -- | Converts a transaction output from the chain index to the plutus-ledger-api
 -- transaction output.
 --
--- Note that converting from 'ChainIndexTxOut' to 'TxOut' and back to
--- 'ChainIndexTxOut' loses precision ('Datum' and 'Validator' are changed to 'DatumHash' and 'ValidatorHash' respectively)
-toTxOut :: ChainIndexTxOut -> TxOut
-toTxOut (PublicKeyChainIndexTxOut addr v)          = TxOut addr v Nothing
-toTxOut (ScriptChainIndexTxOut addr _ (Left dh) v) = TxOut addr v (Just dh)
-toTxOut (ScriptChainIndexTxOut addr _ (Right d) v) = TxOut addr v (Just $ datumHash d)
+-- Note that 'ChainIndexTxOut' supports features such inline datums and
+-- reference scripts which are not supported by V1 TxOut. Converting from
+-- 'ChainIndexTxOut' to 'TxOut' and back is therefore lossy.
+toTxOut :: ChainIndexTxOut -> V1.Tx.TxOut
+toTxOut (PublicKeyChainIndexTxOut addr v _outputDatum _referenceScript) =
+  V1.Tx.TxOut addr v Nothing
+toTxOut (ScriptChainIndexTxOut addr v (Left dh) _referenceScript _validator)     =
+  V1.Tx.TxOut addr v (Just dh)
+toTxOut (ScriptChainIndexTxOut addr v (Right d) _referenceScript _validator)     =
+  V1.Tx.TxOut addr v (Just $ datumHash d)
 
 -- | Converts a plutus-ledger-api transaction output to the chain index
 -- transaction output.
-fromTxOut :: TxOut -> Maybe ChainIndexTxOut
-fromTxOut TxOut { txOutAddress, txOutValue, txOutDatumHash } =
-  case addressCredential txOutAddress of
-    PubKeyCredential _ -> pure $ PublicKeyChainIndexTxOut txOutAddress txOutValue
-    ScriptCredential vh ->
+fromTxOut :: V1.Tx.TxOut -> Maybe ChainIndexTxOut
```

Comment:
May I suggest testing the  _partial_ inverse property test between `toTxOut` and `fromTxOut`? In the test, you could generate a `ChainIndexTxOut`, convert it to a `TxOut` and back to `ChainIndexTxOut`. Then you can verify what the expected result would be.

Reply:
> In the test, you could generate a `ChainIndexTxOut`, convert it to a `TxOut` and back to `ChainIndexTxOut`. Then you can verify what the expected result would be.

This is wrong. From `ChainIndexTxOut` to `TxOut` is lossy as `ChainIndexTxOut` can have the full datum while `TxOut` only its hash. You probably meant the other way around `TxOut` -> `ChainIndex` -> `TxOut` and we have already a test for that. https://github.com/input-output-hk/plutus-apps/blob/6d1600f084e894a257e02d3d0095b1f7a752b464/plutus-ledger/test/Spec.hs#L229-L234

Reply:
That's right! I think that's the point, you can also test the lossy conversion to make sure we lose just enough.

Reply:
:thinking:

Reply:
Ultimately, it's up to you :). If you think it doesn't make sense, no need to do it.